### PR TITLE
MultiIndex union/intersection with non-object other

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

ATM if we call MultiIndex.union with e.g. a Float64Index, we get `ValueError: Buffer dtype mismatch, expected 'Python object' but got 'double'` from `lib.fast_unique_multiple`.  This PR changes that to raise NotImplementedError with a reasonable message, but we could alternatively return `self.to_flat_index().union(other)`

I prefer making the user do to_flat_index on their own because this seems like something that would often be reached by accident.

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3235,9 +3235,13 @@ def union(self, other, sort=None):
 
         # TODO: Index.union returns other when `len(self)` is 0.
 
-        uniq_tuples = lib.fast_unique_multiple(
-            [self._values, other._ndarray_values], sort=sort
-        )
+        if not is_object_dtype(other.dtype):
+            raise NotImplementedError(
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: is this hit in any tests?

Comment (jbrockmendel) [reply pandas/core/indexes/multi.py]: no, can add one

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -3235,9 +3235,13 @@ def union(self, other, sort=None):
...
+                "Can only union MultiIndex with MultiIndex or Index of tuples, "
+                "try mi.to_flat_index().union(other) instead."
+            )
+
+        uniq_tuples = lib.fast_unique_multiple([self._values, other._values], sort=sort)
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/multi.py]: By not passing an ndarray (`_ndarray_values`), I think `fast_unique_multiple` might get a lot slower?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/multi.py]: we only get here with object dtype, so they are equivalent

Comment (jorisvandenbossche) [reply pandas/core/indexes/multi.py]: Ah, yes, I had interpreted the `if not is_object_dtype(..)` the other way around ..

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/multi.py]: can you add one

Comment (jbrockmendel) [reply pandas/core/indexes/multi.py]: added+green (though no azure)

**Comment by jbrockmendel** [comment]
gentle ping; we're on the verge of being rid of _ndarray_values

**Comment by jreback** [comment]
thanks

**Pull request merged**